calcom/cal.com
# Team Billing

## What does this PR do?
Charge team owners on a per seat basis for all non-pro members of a team.

Issue: https://github.com/calendso/calendso/issues/1472

### How to review
- Set `HOSTED_CAL_FEATURES=true` in your env
- Ensure your env is authed with Cal's official Stripe testing account
- Sign into pro@example.com and create a team
- Add members, remove members, leave teams & witness billing update accordingly

## Test Cases
- Team with existing members without seats shows alert notice for team owner
- Team owner can purchase missing seats
- Members are upgraded to pro when added to a team
- Seat is added in background when member is added
- Members with Pro are not counted when purchasing seats
- Owner is not counted when purchasing seat
- Members on trial status are included in seat count
- Members on free status are hidden from team
- Members are downgraded when team is deleted
- Members are downgraded when removed from/leave team
- Members with their own team can purchase seats for their members while their pro sponsored by another team
- Members with their own team are prompted to upgrade to pro when removed from team sponsoring their pro

## Deployment Strategy
- Email detailing new billing system sent to all Pro members that have a team
```sql
SELECT username, email
FROM users
LEFT JOIN "Membership" m ON m."userId" = id
WHERE role = 'OWNER'::"MembershipRole"
```
- All team members (not owner) without a stripe subscription are downgraded to Trial status
```sql
UPDATE users u	
SET plan = 'TRIAL'::"UserPlan"
FROM (
    SELECT id, metadata
    FROM users
        LEFT JOIN "Membership" m ON m."userId" = id
    WHERE plan = 'PRO'::"UserPlan"
	AND metadata IS NULL
	AND m."role" != 'OWNER'::"MembershipRole"
) as usrs
WHERE  u.id = usrs.id
```
- ~Flexible pro pricing options to be created on Stripe~ ([done](https://dashboard.stripe.com/prices/price_1KHkoeH8UDiwIftkkUbiggsM))
- Ensure multiple owners can't exist for one team (check prod for legacy teams) 

### Todo:
- [x] [Backend] Seat calculator: `getTeamSeatStats` .
        - Returns object of keys `unpaidSeats` and `freeSeats`.
        - Free seats are the owner themselves and any existing pro accounts.
- [x] [Backend] Upgrade function: `upgradeTeam`.
        - Handles switching to flexible pro pricing option with quantity derived from `unpaidSeats`.
- [x] [Backend] Downgrade function: `downgradeTeamMembers`.
        - Fired when Stripe reports failed billing for team owner's pro account
- [x] Notice + CTA to purchase missing seats for existing members.
        - Placed on team settings page ~and/or at bottom of sidebar~.
- [x] Notice when adding member that it will increase the team's seat count.
        - A modal with information about how the seat count will be increased and CTA to continue.
- [x] Increase Stripe subscription quantity when new member is added.
- [x] Decrease Stripe subscription quantity when a member is removed.
- [x] A notice that the team owner must upgrade, or that they can purchase a pro license.
        - Team owner has 14 days to upgrade while team members are switched to trial plan.
        - ~Potentially notice on event type page.~
- [x] Badge with tooltip on team list.
- [x] Hide free users from public team page
- [x] Remove "Create Event Type" button for non admins
- [x] Fix padding issue with availability button on MemberListItem
- [x] HOSTED_CAL_FEATURES to enable team billing for testing, defaults to TRUE in prod 

Billing could be discounted at scale, for example:
- $12 per month per user for first 10 users
- $10 per month per user for each additional user for first 100 users
- $8 per month per user for each additional user
This however will be configurable via Stripe depending on the business department. 

## Type of change


## Checklist:
